tiledb/query_condition.py

```diff
@@ -100,17 +117,48 @@ def visit_Compare(self, node):
             raise tiledb.TileDBError("Incorrect type for attribute name.")
 
         if isinstance(val, ast.Constant):
-            val = val.value
+            sign = val.sign if hasattr(val, "sign") else 1
+            val = val.value * sign
         elif isinstance(val, ast.Num):
             # deprecated in 3.8
-            val = val.n
+            sign = val.sign if hasattr(val, "sign") else 1
+            val = val.n * sign
         elif isinstance(val, ast.Str) or isinstance(val, ast.Bytes):
             # deprecated in 3.8
             val = val.s
         else:
-            raise tiledb.TileDBError("Incorrect type for comparison value.")
+            raise tiledb.TileDBError(
+                f"Incorrect type for comparison value: {ast.dump(val)}"
+            )
+
+        if not self._schema.has_attr(att):
+            raise tiledb.TileDBError(f"Attribute `{att}` found not in schema.")
+
+        if att not in self._query_attrs:
+            raise tiledb.TileDBError(
+                f"Attribute `{att}` given to filter in query's `attr_cond` "
+                "arg but not found in `attr` arg."
+            )
+
+        dtype = self._schema.attr(att).dtype
+
+        if dtype.kind in "SUa":
+            dtype_name = "string"
+        else:
+            try:
+                cast = getattr(np, dtype.name)
+                val = cast(val)
+                dtype_name = dtype.name
+            except ValueError:
+                raise tiledb.TileDBError(
+                    f"Type mismatch between attribute `{att}` and value `{val}`."
+                )
 
-        return qc.qc(att, val, op, self._ctx)
+        c_obj = qc.qc(self._ctx)
+        init_qc = getattr(c_obj, f"init_{dtype_name}")
```

Comment:
What happens when:
- the dtype isn't supported -- we get an `AttributeError`, right? IMO that's going to be a little confusing so probably need a try/except here to explain the error
- the supplied value is out of bounds for the dtype: we get a `ValueError` or a `TypeError` for invalid cast? Probably also could use some error clarification.

Reply:
Good points. I will add error checks for these then.

Reply:
For your second question, an out of bounds value will throw a `tiledb.libtiledb.TileDBError: OverflowError: Python int too large to convert to C long` from the C++ end.